Enabling --combined-json to output to file using the -o flag

ali92hm opened the pull request:
Prior to these changes `solc --combined-json abi,bin Contract.sol -o outdir` ignored the output dir and wrote the output to stdout. 

This PR enables you to specify the output dir for the JSON file to be written to.

ali92hm committed:
Added createJson method to make writing json easier

ali92hm committed:
Enabling --combined-json to output to file

chriseth commented on solc/CommandLineInterface.cpp:
```diff
@@ -843,7 +849,11 @@ void CommandLineInterface::handleCombinedJSON()
-	cout << dev::jsonCompactPrint(output) << endl;
+
+	if (m_args.count(g_argOutputDir))
+		createJson(m_args[g_argInputFile].as<vector<string>>().front(), output);
```
Why did you decide to name the file after one of the input files? The combined json should contain the compilation result from all input files. What about `result.json` a filename?

ali92hm replied:
Yes, that makes perfect sense. Either `result.json` or `combined.json`

axic commented on solc/CommandLineInterface.cpp:
```diff
@@ -486,6 +486,12 @@ void CommandLineInterface::createFile(string const& _fileName, string const& _da
 		BOOST_THROW_EXCEPTION(FileError() << errinfo_comment("Could not write to file: " + pathName));
 }
 
+void CommandLineInterface::createJson(string const& _fileName, Json::Value const& _json)
```
I think this would be generic having a flag to switch between `jsonCompactPrint` and `jsonPrettyPrint` given some outputs are pretty printed.

Perhaps as a feature that could be considered to pretty print all combined output.

ali92hm replied:
Done. Should there be a command line flag to switch between the two?

ali92hm commented:
@chriseth Changed the output file name to `result.json`

@axic Added a flag for switching between `jsonCompactPrint` and `jsonPrettyPrint`. I feel like there should be command line flag to let the user decide.

ali92hm commented:
Does it make sense to change the [line](https://github.com/ethereum/solidity/blob/develop/solc/CommandLineInterface.cpp#L282)
`createFile(m_compiler->filesystemFriendlyName(_contract) + ".abi", data);` to `createJson(m_compiler->filesystemFriendlyName(_contract) + ".abi", data);`

This way the name of ABI will be `ContractName.abi.json` and printed in pretty format by default.

ali92hm commented on solc/CommandLineInterface.cpp:
```diff
@@ -790,7 +799,7 @@ void CommandLineInterface::handleCombinedJSON()
 	{
 		Json::Value contractData(Json::objectValue);
 		if (requests.count(g_strAbi))
-			contractData[g_strAbi] = dev::jsonCompactPrint(m_compiler->contractABI(contractName));
```
JSON should be rendered to string when its being printed out and not here.

axic replied:
While I like this way better, there might be others using the combined JSON as is and this is a breaking change for them. Probably we should do it separately, or rather encourage others to use the *standard json* interface instead.

ali92hm replied:
Sounds good, I reverted those.

axic replied:
I still think this and the natspec fields cannot be changed from serialised in a regular release, though it should be  fine adding the pretty print as an option here. We're aiming to another breaking release (0.5.0) soon, which could then include this part of the change.

What do you think?

ali92hm replied:
Sounds good. I was gonna make another change. The output of `solc --abi` is a file called `contractName.abi`

What do you think about changing that to `contractName.abi.json` and give the option to `--pretty-json` print it or compact?

axic replied:
Can you split that discussion into a new PR?

Please remove change to these three fields (abi, natspecdev, natspecuser) so this PR remains non-breaking.

ali92hm replied:
I removed the change for `--abi` but the whole point of this PR was to make change to `abi, natspecdev, natspecuser`  and give the user the option to use `--pretty-json` if they dont want compact.

If `--pretty-json` is not passed the output should not be different from before. Isn't that why we are shooting for `0.5.0` release?

axic replied:
Currently, the "compact json" has three members (abi, natspecdev, natspecuser) which are already serialised to strings, even before the "compact json" is.

I think we can safely introduce the option to pretty printing to the outside JSON, but shouldn't change those three fields as tools could expect them to be strings.

chriseth commented:
@Ali92hm thanks for the changes! I think I would prefer `combined.json` (as you suggested) over `result.json`. For the rest of the changes, I would pass to @axic.

ali92hm commented:
@chriseth I changed the output name to `combined.json`.

@axic I reverted the `dev::jsonCompactPrint` stuff. 

Let me know if there is anything else you want me to change!

axic commented on solc/CommandLineInterface.cpp:
```diff
@@ -507,6 +507,15 @@ void CommandLineInterface::createFile(string const& _fileName, string const& _da
 		BOOST_THROW_EXCEPTION(FileError() << errinfo_comment("Could not write to file: " + pathName));
 }
 
+void CommandLineInterface::createJson(string const& _fileName, Json::Value const& _json, bool pretty)
```
I think it would make sense naming this `createJsonFile`

axic commented on solc/CommandLineInterface.cpp:
```diff
@@ -894,7 +903,11 @@ void CommandLineInterface::handleCombinedJSON()
+	if (m_args.count(g_argOutputDir))
+		createJson("combined", output);
+	else
+		cout << dev::jsonPrettyPrint(output) << endl;
```
This changes behaviour, from compact to pretty printed.

axic commented on solc/CommandLineInterface.h:
```diff
@@ -81,6 +81,12 @@ class CommandLineInterface
+	/// @arg _fileName the name of the file (the extension will be replaced with .json)
+	/// @arg _json to be written
+	/// @arg pretty boolean indicating the json to be written in compact or pretty form (true by default)
+	void createJson(std::string const& _fileName, Json::Value const& _json, bool pretty=true);
```
Please add spaces around `=`.

This also changes behaviour.

Perhaps it would make more sense adding a flag, `--prettyprint` to output pretty printed JSON?

ali92hm replied:
Yup, thats what I was asking. Will do

ali92hm replied:
If I'm doing that should I get rid of `dev::jsonCompactPrint` in the intermediary JSON object and decide on the `jsonPrettyPrint` or `jsonCompactPrint` at the time of the output?
```
		if (requests.count(g_strAbi))
			contractData[g_strAbi] = dev::jsonCompactPrint(m_compiler->contractABI(contractName));
```

axic replied:
I think that could be fine and shouldn't not break the output for others.

ali92hm replied:
Yeah, I agree. Will be updating the PR soon.
Also any ideas why the Appveyor CI fails?

axic replied:
You can ignore that Appveyor failure. It is a problem with pull requests from remote repos we haven't addressed yet.

axic commented:
@Ali92hm would you have some time finishing this? It would be a nice addition to be merged.

ali92hm committed:
Added --pretty-json commandline option

ali92hm commented:
@axic I made some changes and added `--pretty-json` commandline option.

Let me know what you think.

axic commented:
@chriseth rebased and changed, please review

chriseth commented on solc/CommandLineInterface.cpp:
```diff
@@ -541,6 +548,7 @@ Allowed options)",
 			"Estimated number of contract runs for optimizer tuning."
 		)
 		(g_argAddStandard.c_str(), "Add standard contracts.")
+		(g_argPrettyJson.c_str(), "Output JSON in pretty format.")
```
It is a little confusing that this only applies to `--combined-json` and not to `--standard-json`.

axic replied:
Hm, actually that is not controlled by the CLI, but is built in standarcompiler.

chriseth replied:
Ok, but `--help` should explain that.

axic replied:
We could change the API of StandardCompiler to support that if needed.

axic replied:
Okay changed it for now, let's implement support for StandardCompiler separately.

ali92hm replied:
Is there anything you want me to do?

axic replied:
Thanks for the contribution, it is now merged :)
